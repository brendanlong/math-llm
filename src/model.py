"""Small transformer model for arithmetic tasks.

This module implements a decoder-only transformer model designed for
learning basic arithmetic operations like addition.
"""

import math
from typing import Any, Optional

import torch
import torch.nn as nn
from torch.nn import functional as F


class PositionalEncoding(nn.Module):
    """Sinusoidal positional encoding for transformer inputs."""

    pe: torch.Tensor

    def __init__(self, d_model: int, max_len: int = 32):
        """Initialize positional encoding.

        Args:
            d_model: Model dimension
            max_len: Maximum sequence length
        """
        super().__init__()

        pe = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
        div_term = torch.exp(
            torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model)
        )

        pe[:, 0::2] = torch.sin(position * div_term)
        pe[:, 1::2] = torch.cos(position * div_term)
        pe = pe.unsqueeze(0).transpose(0, 1)

        self.register_buffer("pe", pe)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        """Add positional encoding to input embeddings.

        Args:
            x: Input tensor of shape (seq_len, batch_size, d_model)

        Returns:
            Tensor with positional encoding added
        """
        pe_slice = self.pe[: x.size(0), :]
        return x + pe_slice


class TransformerBlock(nn.Module):
    """Single transformer decoder block with masked self-attention."""

    def __init__(self, d_model: int, n_heads: int, d_ff: int, dropout: float = 0.1):
        """Initialize transformer block.

        Args:
            d_model: Model dimension
            n_heads: Number of attention heads
            d_ff: Feed-forward dimension
            dropout: Dropout probability
        """
        super().__init__()

        self.self_attn = nn.MultiheadAttention(
            d_model, n_heads, dropout=dropout, batch_first=True
        )
        self.feed_forward = nn.Sequential(
            nn.Linear(d_model, d_ff),
            nn.ReLU(),
            nn.Linear(d_ff, d_model),
        )

        self.norm1 = nn.LayerNorm(d_model)
        self.norm2 = nn.LayerNorm(d_model)
        self.dropout = nn.Dropout(dropout)

    def forward(
        self, x: torch.Tensor, mask: Optional[torch.Tensor] = None
    ) -> torch.Tensor:
        """Forward pass through transformer block.

        Args:
            x: Input tensor of shape (batch_size, seq_len, d_model)
            mask: Optional attention mask

        Returns:
            Output tensor of same shape as input
        """
        # Self-attention with residual connection
        attn_out, _ = self.self_attn(x, x, x, attn_mask=mask)
        x = self.norm1(x + self.dropout(attn_out))

        # Feed-forward with residual connection
        ff_out = self.feed_forward(x)
        x = self.norm2(x + self.dropout(ff_out))

        return x


class ArithmeticModel(nn.Module):
    """Small transformer model for arithmetic tasks."""

    def __init__(
        self,
        vocab_size: int = 13,
        d_model: int = 256,
        n_layers: int = 4,
        n_heads: int = 4,
        d_ff: int = 1024,
        max_seq_len: int = 32,
        dropout: float = 0.1,
    ):
        """Initialize the arithmetic model.

        Args:
            vocab_size: Size of vocabulary (default 13 for arithmetic)
            d_model: Model dimension
            n_layers: Number of transformer layers
            n_heads: Number of attention heads
            d_ff: Feed-forward dimension
            max_seq_len: Maximum sequence length
            dropout: Dropout probability
        """
        super().__init__()

        self.d_model = d_model
        self.max_seq_len = max_seq_len

        # Embedding layers
        self.token_embedding = nn.Embedding(vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(d_model, max_seq_len)

        # Transformer layers
        self.layers = nn.ModuleList(
            [TransformerBlock(d_model, n_heads, d_ff, dropout) for _ in range(n_layers)]
        )

        # Output projection
        self.ln_f = nn.LayerNorm(d_model)
        self.lm_head = nn.Linear(d_model, vocab_size, bias=False)

        # Initialize weights
        self.apply(self._init_weights)

    def _init_weights(self, module: nn.Module) -> None:
        """Initialize model weights."""
        if isinstance(module, nn.Linear):
            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
            bias = getattr(module, "bias", None)
            if bias is not None:
                torch.nn.init.zeros_(bias)
        elif isinstance(module, nn.Embedding):
            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
        elif isinstance(module, nn.LayerNorm):
            torch.nn.init.zeros_(module.bias)
            torch.nn.init.ones_(module.weight)

    def _get_causal_mask(self, seq_len: int) -> torch.Tensor:
        """Create causal (lower triangular) attention mask.

        Args:
            seq_len: Sequence length

        Returns:
            Causal mask tensor
        """
        mask = torch.triu(torch.ones(seq_len, seq_len), diagonal=1)
        mask = mask.masked_fill(mask == 1, float("-inf"))
        return mask

    def forward(
        self,
        input_ids: torch.Tensor,
        attention_mask: Optional[torch.Tensor] = None,
        labels: Optional[torch.Tensor] = None,
        **kwargs: Any,
    ) -> dict[str, torch.Tensor] | torch.Tensor:
        """Forward pass through the model.

        Args:
            input_ids: Input token IDs of shape (batch_size, seq_len)
            attention_mask: Optional attention mask
            labels: Optional labels for computing loss
            **kwargs: Additional arguments (ignored)

        Returns:
            If labels provided: dict with 'loss' and 'logits'
            Otherwise: logits tensor of shape (batch_size, seq_len, vocab_size)
        """
        _, seq_len = input_ids.shape

        # Token embeddings
        x = self.token_embedding(input_ids)  # (batch_size, seq_len, d_model)
        x = x * math.sqrt(self.d_model)  # Scale embeddings

        # Positional encoding (convert to batch_first format)
        x = x.transpose(0, 1)  # (seq_len, batch_size, d_model)
        x = self.pos_encoding(x)
        x = x.transpose(0, 1)  # (batch_size, seq_len, d_model)

        # Create causal mask
        causal_mask = self._get_causal_mask(seq_len).to(input_ids.device)

        # Apply transformer layers
        for layer in self.layers:
            x = layer(x, mask=causal_mask)

        # Final layer norm and projection
        x = self.ln_f(x)
        logits = self.lm_head(x)  # (batch_size, seq_len, vocab_size)

        # Compute loss if labels are provided
        if labels is not None:
            # Shift logits and labels for next-token prediction
            shift_logits = logits[..., :-1, :].contiguous()
            shift_labels = labels[..., 1:].contiguous()

            # Flatten for loss computation
            loss_fct = nn.CrossEntropyLoss()
            shift_logits = shift_logits.view(-1, shift_logits.size(-1))
            shift_labels = shift_labels.view(-1)

            loss = loss_fct(shift_logits, shift_labels)
            return {"loss": loss, "logits": logits}

        return logits

    def generate(
        self,
        input_ids: torch.Tensor,
        max_new_tokens: int = 20,
        temperature: float = 1.0,
        top_k: Optional[int] = None,
        end_token_id: int = 12,
    ) -> torch.Tensor:
        """Generate tokens autoregressively.

        Args:
            input_ids: Initial input tokens of shape (batch_size, seq_len)
            max_new_tokens: Maximum number of new tokens to generate
            temperature: Sampling temperature
            top_k: Optional top-k sampling
            end_token_id: Token ID for end-of-sequence

        Returns:
            Generated tokens of shape (batch_size, seq_len + num_generated)
        """
        self.eval()

        for _ in range(max_new_tokens):
            # Get predictions for current sequence
            with torch.no_grad():
                outputs = self.forward(input_ids)
                # Extract logits (forward returns dict when labels provided, tensor otherwise)
                if isinstance(outputs, dict):
                    logits = outputs["logits"]
                else:
                    logits = outputs

                # Get logits for last token
                logits = logits[:, -1, :] / temperature

                # Apply top-k filtering if specified
                if top_k is not None:
                    v, _ = torch.topk(logits, min(top_k, logits.size(-1)))
                    logits[logits < v[:, [-1]]] = -float("inf")

                # Sample next token
                probs = F.softmax(logits, dim=-1)
                next_token = torch.multinomial(probs, num_samples=1)

                # Append to sequence
                input_ids = torch.cat([input_ids, next_token], dim=1)

                # Stop if end token is generated
                if next_token.item() == end_token_id:
                    break

        return input_ids

    def count_parameters(self) -> int:
        """Count total number of trainable parameters."""
        return sum(p.numel() for p in self.parameters() if p.requires_grad)


def create_small_model() -> ArithmeticModel:
    """Create a small model configuration (~1M parameters)."""
    return ArithmeticModel(
        vocab_size=13,
        d_model=256,
        n_layers=4,
        n_heads=4,
        d_ff=512,
        max_seq_len=32,
        dropout=0.1,
    )


def create_medium_model() -> ArithmeticModel:
    """Create a medium model configuration (~5M parameters)."""
    return ArithmeticModel(
        vocab_size=13,
        d_model=512,
        n_layers=6,
        n_heads=8,
        d_ff=1024,
        max_seq_len=32,
        dropout=0.1,
    )


def create_large_model() -> ArithmeticModel:
    """Create a large model configuration (~10M parameters)."""
    return ArithmeticModel(
        vocab_size=13,
        d_model=512,
        n_layers=8,
        n_heads=8,
        d_ff=2048,
        max_seq_len=32,
        dropout=0.1,
    )
