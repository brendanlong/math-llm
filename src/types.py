"""Shared type definitions for the math LLM project."""

from typing import NotRequired, TypedDict


class DatasetMetadata(TypedDict):
    """Metadata for dataset generation."""

    max_digits: int
    max_operands: int
    seed: int
    num_workers: int
    include_chain_of_thought: bool
    fixed_length_cot: bool
    num_examples: int
    split_ratios: dict[str, float]
    generation_timestamp: str
    generation_version: str


class DatasetDict(TypedDict):
    """Type definition for the complete dataset."""

    examples: list[str]
    metadata: NotRequired[DatasetMetadata]
