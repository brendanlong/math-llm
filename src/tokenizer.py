"""Custom character-level tokenizer for arithmetic expressions with reasoning.

This tokenizer handles a vocabulary of 19 tokens:
- Digits: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
- Operators: +, =
- Special: <end>, <noop>
- Multi-operand reasoning: <think_multi>, </think_multi>
- Multi-digit reasoning: <think_digit>, </think_digit>
- Formatting: \n (newline)
"""

import tempfile

from tokenizers import (
    Regex,
    Tokenizer,
    decoders,
    models,
    pre_tokenizers,
)
from transformers.tokenization_utils_fast import PreTrainedTokenizerFast

# Vocabulary mapping for arithmetic expressions with reasoning
VOCAB = {
    "0": 0,
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "+": 10,
    "=": 11,
    "<end>": 12,
    "<think_multi>": 13,
    "</think_multi>": 14,
    "<think_digit>": 15,
    "</think_digit>": 16,
    "\n": 17,
    "<noop>": 18,
}

TOKEN_PATTERN = (
    r"</think_multi>|</think_digit>|<think_multi>|<think_digit>|<noop>|<end>|\n|[0-9+=]"
)

# Constants derived from vocabulary
VOCAB_SIZE = len(VOCAB)

tokenizer = Tokenizer(models.WordLevel(vocab=VOCAB, unk_token=None))
tokenizer.pre_tokenizer = pre_tokenizers.Split(
    pattern=Regex(TOKEN_PATTERN), behavior="isolated"
)  # type: ignore
tokenizer.decoder = decoders.Fuse()  # type: ignore

end_token_id = VOCAB["<end>"]
with tempfile.NamedTemporaryFile() as f:
    tokenizer.save(f.name)

    fast_tokenizer = PreTrainedTokenizerFast(
        tokenizer_file=f.name,
        unk_token=None,
        pad_token="<end>",
        eos_token="<end>",
        clean_up_tokenization_spaces=False,
    )


class ArithmeticTokenizer:
    """Character-level tokenizer for arithmetic expressions."""

    vocab = VOCAB

    def __init__(self):
        """Initialize the tokenizer with arithmetic vocabulary."""
        self.tokenizer = fast_tokenizer

        self.end_token_id = VOCAB["<end>"]

    @property
    def vocab_size(self) -> int:
        """Return the vocabulary size."""
        return VOCAB_SIZE

    def encode(self, text: str) -> list[int]:
        """Encode text to token IDs.

        Args:
            text: Input text (e.g., "3+5=8<end>")

        Returns:
            List of token IDs

        Raises:
            Exception: If text contains unknown characters
        """
        return self.tokenizer.encode(text)

    def decode(self, token_ids: list[int]) -> str:
        """Decode token ID(s) to text.

        Args:
            token_ids: List of token IDs

        Returns:
            Decoded text string
        """
        return self.tokenizer.decode(token_ids)
