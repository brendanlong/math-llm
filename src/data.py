"""Data loading utilities for arithmetic expression datasets."""

import json
from pathlib import Path

import torch
from torch.utils.data import DataLoader, Dataset

from .model import MAX_SEQUENCE_LENGTH
from .tokenizer import ArithmeticTokenizer
from .types import DatasetDict


class ArithmeticDataset(Dataset[dict[str, torch.Tensor]]):
    """Dataset for arithmetic expressions.

    Loads expressions from JSON files and tokenizes them for training.
    """

    def __init__(
        self,
        data_path: str | Path,
        tokenizer: ArithmeticTokenizer,
        max_length: int = MAX_SEQUENCE_LENGTH,
    ):
        """Initialize dataset.

        Args:
            data_path: Path to JSON file containing expressions
            tokenizer: Tokenizer instance for encoding expressions
            max_length: Maximum sequence length (sequences will be padded/truncated)
        """
        self.data_path = Path(data_path)
        self.tokenizer = tokenizer
        self.max_length = max_length

        # Load data
        with open(self.data_path, "r") as f:
            dataset: DatasetDict = json.load(f)
            self.data = dataset["examples"]

    def __len__(self) -> int:
        return len(self.data)

    def __getitem__(self, idx: int) -> dict[str, torch.Tensor]:
        """Get a single tokenized example.

        Args:
            idx: Index of the example

        Returns:
            Dictionary with 'input_ids' and 'labels' tensors
        """
        expression = self.data[idx]["text"]

        # Split at equals sign for completion-style training
        if "=" in expression:
            parts = expression.split("=", 1)
            prompt = parts[0] + "="  # e.g., "3+5="
            completion = parts[1]  # e.g., "8<end>" or "<think>...</think>8<end>"
        else:
            # Fallback for malformed data
            prompt = expression
            completion = ""

        # Tokenize prompt and completion separately
        prompt_tokens = self.tokenizer.encode(prompt)
        completion_tokens = self.tokenizer.encode(completion) if completion else []

        # Combine for full sequence
        full_tokens = prompt_tokens + completion_tokens

        # Pad or truncate to max_length
        if len(full_tokens) > self.max_length:
            full_tokens = full_tokens[: self.max_length]
            # Adjust prompt length if needed
            if len(prompt_tokens) >= self.max_length:
                prompt_tokens = prompt_tokens[: self.max_length - 1]
        else:
            # Pad with end token
            pad_length = self.max_length - len(full_tokens)
            full_tokens = full_tokens + [self.tokenizer.vocab["<end>"]] * pad_length

        # Convert to tensor
        input_ids = torch.tensor(full_tokens, dtype=torch.long)

        # For completion-style training:
        # - Input is the full sequence
        # - Labels mask the prompt part, only train on completion
        labels = input_ids.clone()

        # Mask the prompt tokens in labels (don't compute loss on them)
        prompt_length = len(prompt_tokens)
        for i in range(prompt_length):
            labels[i] = -100

        # Mask padding tokens in labels
        original_length = len(prompt_tokens) + len(completion_tokens)
        if original_length < self.max_length:
            for i in range(original_length, self.max_length):
                labels[i] = -100

        return {"input_ids": input_ids, "labels": labels}


def create_dataloader(
    data_path: str | Path,
    tokenizer: ArithmeticTokenizer,
    batch_size: int = 32,
    shuffle: bool = True,
    max_length: int = MAX_SEQUENCE_LENGTH,
    num_workers: int = 0,
) -> DataLoader[dict[str, torch.Tensor]]:
    """Create a DataLoader for arithmetic expressions.

    Args:
        data_path: Path to JSON file containing expressions
        tokenizer: Tokenizer instance
        batch_size: Batch size for training
        shuffle: Whether to shuffle the data
        max_length: Maximum sequence length
        num_workers: Number of worker processes for data loading

    Returns:
        DataLoader instance ready for training/evaluation
    """
    dataset = ArithmeticDataset(
        data_path=data_path, tokenizer=tokenizer, max_length=max_length
    )

    return DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=shuffle,
        num_workers=num_workers,
        pin_memory=torch.cuda.is_available(),
    )


def load_splits(
    data_dir: str | Path,
    tokenizer: ArithmeticTokenizer,
    batch_size: int = 32,
    max_length: int = MAX_SEQUENCE_LENGTH,
    num_workers: int = 0,
) -> tuple[
    DataLoader[dict[str, torch.Tensor]],
    DataLoader[dict[str, torch.Tensor]],
    DataLoader[dict[str, torch.Tensor]],
]:
    """Load train, validation, and test DataLoaders.

    Args:
        data_dir: Directory containing train.json, val.json, test.json
        tokenizer: Tokenizer instance
        batch_size: Batch size for all splits
        max_length: Maximum sequence length
        num_workers: Number of worker processes

    Returns:
        Tuple of (train_loader, val_loader, test_loader)
    """
    data_dir = Path(data_dir)

    train_loader = create_dataloader(
        data_path=data_dir / "train.json",
        tokenizer=tokenizer,
        batch_size=batch_size,
        shuffle=True,
        max_length=max_length,
        num_workers=num_workers,
    )

    val_loader = create_dataloader(
        data_path=data_dir / "val.json",
        tokenizer=tokenizer,
        batch_size=batch_size,
        shuffle=False,
        max_length=max_length,
        num_workers=num_workers,
    )

    test_loader = create_dataloader(
        data_path=data_dir / "test.json",
        tokenizer=tokenizer,
        batch_size=batch_size,
        shuffle=False,
        max_length=max_length,
        num_workers=num_workers,
    )

    return train_loader, val_loader, test_loader
