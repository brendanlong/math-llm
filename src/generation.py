"""Data generation utilities for arithmetic expressions with chain-of-thought reasoning."""

import random

from .tokenizer import ArithmeticTokenizer


def generate_two_operand_chain_of_thought(a: int, b: int) -> str:
    """Generate chain-of-thought reasoning for addition.

    Args:
        a: First operand
        b: Second operand

    Returns:
        Chain-of-thought reasoning string with multi-digit thinking tokens
    """
    str_a = str(a)
    str_b = str(b)

    # For single-digit addition, no reasoning needed
    if len(str_a) == 1 and len(str_b) == 1:
        return ""

    # Pad numbers to same length for column addition
    max_len = max(len(str_a), len(str_b))
    str_a = str_a.zfill(max_len)
    str_b = str_b.zfill(max_len)

    reasoning = ["<think_digit>"]
    carry = 0

    # Work from right to left
    for i in range(max_len - 1, -1, -1):
        digit_a = int(str_a[i])
        digit_b = int(str_b[i])

        # Add digits plus any carry
        sum_digits = digit_a + digit_b + carry

        # Show the addition step with recursive thinking when there's a carry (3 numbers)
        if carry > 0:
            reasoning.append(f"\n{digit_a}+{digit_b}+{carry}=")
            # Use recursive thinking for three-number addition
            recursive_reasoning = generate_chain_of_thought([digit_a, digit_b, carry])
            reasoning.append(recursive_reasoning)
            reasoning.append(str(sum_digits))
        else:
            reasoning.append(f"\n{digit_a}+{digit_b}={sum_digits}")

        # Update carry for next iteration
        carry = sum_digits // 10 if sum_digits >= 10 else 0

    reasoning.append("</think_digit>")
    return "".join(reasoning)


def generate_chain_of_thought(operands: list[int]) -> str:
    """Generate recursive left-to-right chain-of-thought for multiple operands.

    Args:
        operands: List of operands to add (e.g., [3, 5, 2])

    Returns:
        Chain-of-thought reasoning string showing recursive addition with nested thinking tags
    """
    if len(operands) < 3:
        # For 2 operands, use original chain-of-thought
        return generate_two_operand_chain_of_thought(operands[0], operands[1])

    # For 3+ operands, show recursive left-to-right addition
    reasoning = ["<think_multi>"]

    # Start with first operand
    current_sum = operands[0]

    for i in range(1, len(operands)):
        next_operand = operands[i]

        # Show the addition step
        reasoning.append(f"\n{current_sum}+{next_operand}=")

        # Get the reasoning for this step (with nested thinking tags if multi-digit)
        step_reasoning = generate_two_operand_chain_of_thought(
            current_sum, next_operand
        )
        if step_reasoning:
            # Keep the nested thinking tags for recursive reasoning
            reasoning.append(step_reasoning)

        # Calculate and show the result
        current_sum = current_sum + next_operand
        reasoning.append(str(current_sum))

    reasoning.append("</think_multi>")
    return "".join(reasoning)


def calculate_max_operand_digits(operands: list[int]) -> int:
    return max(len(str(operand)) for operand in operands)


def generate_addition_examples(
    num_examples: int,
    max_digits: int = 3,
    seed: int = 42,
    max_operands: int = 3,
    fixed_length_cot: bool = False,
) -> list[str]:
    """Generate addition examples with chain-of-thought for multi-digit problems.

    Args:
        num_examples: Number of examples to generate
        max_digits: Maximum number of digits per operand (1-8)
        seed: Random seed for reproducibility
        max_operands: The maximum number of operands to add
        fixed_length_cot: Whether to pad CoT to fixed length with <noop> tokens

    Returns:
        List of arithmetic expressions in format "a+b=<think_digit>...</think_digit>c<end>"
        or "a+b+c=<think_multi>...</think_multi>d<end>" with recursive reasoning
        If fixed_length_cot is True, pads reasoning to 10 * max(digits in operands) * max(number of operands)
    """
    assert max_digits >= 1
    assert max_operands >= 2

    random.seed(seed)
    examples = []
    max_value = 10**max_digits - 1
    tokenizer = ArithmeticTokenizer()
    cot_length = 20 * max_digits * max_operands

    for _ in range(num_examples):
        operands = [
            random.randint(0, max_value) for _ in range(random.randint(2, max_operands))
        ]
        result = sum(operands)

        # Generate chain-of-thought reasoning
        reasoning = generate_chain_of_thought(operands)

        # Apply fixed-length padding if enabled
        if fixed_length_cot:
            max_digits = calculate_max_operand_digits(operands)
            reasoning_length = len(tokenizer.encode(reasoning))
            pad_length = cot_length - reasoning_length
            if pad_length < 0:
                # Error out if our 4x assumption is wrong
                raise ValueError(
                    f"Generated CoT is longer than fixed-length CoT: {reasoning_length} > {cot_length}: {'+'.join(map(str, operands))}={reasoning}"
                )
            elif pad_length > 0:
                reasoning += "<noop>" * pad_length

        example = f"{'+'.join(map(str, operands))}={reasoning}{result}<end>"
        examples.append(example)

    return examples


def split_data(
    examples: list[str], train_ratio: float = 0.8, val_ratio: float = 0.1
) -> tuple[list[str], list[str], list[str]]:
    """Split data into train/validation/test sets.

    Args:
        examples: List of examples to split
        train_ratio: Fraction for training set
        val_ratio: Fraction for validation set (test gets remainder)

    Returns:
        Tuple of (train_examples, val_examples, test_examples)
    """
    total = len(examples)
    train_size = int(total * train_ratio)
    val_size = int(total * val_ratio)

    # Shuffle examples before splitting
    shuffled = examples.copy()
    random.shuffle(shuffled)

    train_examples = shuffled[:train_size]
    val_examples = shuffled[train_size : train_size + val_size]
    test_examples = shuffled[train_size + val_size :]

    return train_examples, val_examples, test_examples
