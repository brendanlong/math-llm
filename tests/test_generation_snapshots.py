"""Snapshot tests for data generation to ensure consistent output."""

from typing import Any

from src.generation import generate_addition_examples


def test_chain_of_thought_snapshots(snapshot: Any) -> None:
    """Test test generation with chain of thought."""
    # Two operands
    assert (
        generate_addition_examples(
            num_examples=2,
            max_digits=1,
            seed=12,
            max_operands=2,
            include_chain_of_thought=True,
        )
        == snapshot
    )

    # Three operands simple
    assert (
        generate_addition_examples(
            num_examples=2,
            max_digits=1,
            seed=13,
            max_operands=3,
            include_chain_of_thought=True,
        )
        == snapshot
    )

    # Three operands with more digits
    assert (
        generate_addition_examples(
            num_examples=2,
            max_digits=3,
            seed=12,
            max_operands=3,
            include_chain_of_thought=True,
        )
        == snapshot
    )

    # Large number of operands and digits
    assert (
        generate_addition_examples(
            num_examples=2,
            max_digits=5,
            seed=12,
            max_operands=5,
            include_chain_of_thought=True,
        )
        == snapshot
    )


def test_no_chain_of_thought_snapshots(snapshot: Any) -> None:
    """Test test generation with no chain of thought."""
    # Two operands
    assert (
        generate_addition_examples(
            num_examples=2,
            max_digits=1,
            seed=12,
            max_operands=2,
            include_chain_of_thought=False,
        )
        == snapshot
    )

    # Three operands simple
    assert (
        generate_addition_examples(
            num_examples=2,
            max_digits=1,
            seed=13,
            max_operands=3,
            include_chain_of_thought=False,
        )
        == snapshot
    )

    # Three operands with more digits
    assert (
        generate_addition_examples(
            num_examples=2,
            max_digits=3,
            seed=12,
            max_operands=3,
            include_chain_of_thought=False,
        )
        == snapshot
    )

    # Large number of operands and digits
    assert (
        generate_addition_examples(
            num_examples=2,
            max_digits=5,
            seed=12,
            max_operands=5,
            include_chain_of_thought=False,
        )
        == snapshot
    )
