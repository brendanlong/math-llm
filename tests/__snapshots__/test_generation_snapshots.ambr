# serializer version: 1
# name: test_chain_of_thought_snapshots
  ''
# ---
# name: test_chain_of_thought_snapshots.1
  '''
  <think_digit>
  2+4=6
  1+3=4</think_digit>
  '''
# ---
# name: test_chain_of_thought_snapshots.2
  '''
  <think_digit>
  8+7=15
  2+1+1=<think_multi>
  2+1=3
  3+1=4</think_multi>4</think_digit>
  '''
# ---
# name: test_chain_of_thought_snapshots.3
  '''
  <think_digit>
  8+9=17
  5+8+1=<think_multi>
  5+8=13
  13+1=<think_digit>
  3+1=4
  1+0=1</think_digit>14</think_multi>14
  6+1+1=<think_multi>
  6+1=7
  7+1=8</think_multi>8</think_digit>
  '''
# ---
# name: test_recursive_chain_of_thought_snapshots
  '''
  <think_digit>
  8+7=15
  2+1+1=<think_multi>
  2+1=3
  3+1=4</think_multi>4</think_digit>
  '''
# ---
# name: test_recursive_chain_of_thought_snapshots.1
  '''
  <think_multi>
  3+5=8
  8+2=10</think_multi>
  '''
# ---
# name: test_recursive_chain_of_thought_snapshots.2
  '''
  <think_multi>
  28+17=<think_digit>
  8+7=15
  2+1+1=<think_multi>
  2+1=3
  3+1=4</think_multi>4</think_digit>45
  45+94=<think_digit>
  5+4=9
  4+9=13</think_digit>139</think_multi>
  '''
# ---
# name: test_recursive_chain_of_thought_snapshots.3
  '''
  <think_multi>
  658+189=<think_digit>
  8+9=17
  5+8+1=<think_multi>
  5+8=13
  13+1=<think_digit>
  3+1=4
  1+0=1</think_digit>14</think_multi>14
  6+1+1=<think_multi>
  6+1=7
  7+1=8</think_multi>8</think_digit>847
  847+234=<think_digit>
  7+4=11
  4+3+1=<think_multi>
  4+3=7
  7+1=8</think_multi>8
  8+2=10</think_digit>1081</think_multi>
  '''
# ---
# name: test_recursive_chain_of_thought_snapshots.4
  '''
  <think_multi>
  658+189=<think_digit>
  8+9=17
  5+8+1=<think_multi>
  5+8=13
  13+1=<think_digit>
  3+1=4
  1+0=1</think_digit>14</think_multi>14
  6+1+1=<think_multi>
  6+1=7
  7+1=8</think_multi>8</think_digit>847
  847+234=<think_digit>
  7+4=11
  4+3+1=<think_multi>
  4+3=7
  7+1=8</think_multi>8
  8+2=10</think_digit>1081
  1081+99=<think_digit>
  1+9=10
  8+9+1=<think_multi>
  8+9=17
  17+1=<think_digit>
  7+1=8
  1+0=1</think_digit>18</think_multi>18
  0+0+1=<think_multi>
  0+0=0
  0+1=1</think_multi>1
  1+0=1</think_digit>1180</think_multi>
  '''
# ---
